### ISSUE TITLE
As an app, I need a way to relay messages back to client

### ISSUE THREAD

**Issue by hshayesteh** [dambileh/Civitas#7]
Depends on #1, #6
We will be making use of socket.it

- Do the necessary research and create documentation that covers all the important topics as far as working with Socket.io is concerned. Here is working example of Redis PubSub and Socket.io in action: https://github.com/rajaraodv/redispubsub

(The socket.io server's implementation is in https://github.com/rajaraodv/redispubsub/blob/master/events/index.js)

- Here is another resource http://chimera.labs.oreilly.com/books/1234000001808/ch07.html#middleware

- Do not worry about the client side stuff too much at this stage, as long as you could establish a connection from the client to the server for testing purposes then it should be enough. There is another ticket that will deal with that side: https://github.com/dambileh/Civitas/issues/8
Here is an example of the code that connects to the socket.io server: https://github.com/rajaraodv/redispubsub/blob/master/public/client.js

## Must have
- Create a WebSocket Adapter that will be shared amongst all the micro services. This allows us to hide the underlying technology from the rest of our application and switch to another if deemed necessary without affecting the internals of our application. Adapter exposes the following functionalities:

     - Listen to messages returned from client i.e **socket.on**
      - Emit messages to client **socket.emit**
      - Handling errors
      - Come up with a naming convention for channels

- Find an answer to this https://github.com/dambileh/Civitas/wiki/Challenges#in-socketio-clients-list-how-do-we-tie-a-client-to-a-user-in-our-application-we-need-that-if-we-want-to-send-targeted-multi-or-uni-cast-messages. This might be the answer: https://stackoverflow.com/questions/10609564/socketio-mysql-authentication/10622204#10622204 . we might be able to pass the UserId back from the client.

- Keep this in mind https://stackoverflow.com/questions/5739357/how-to-reuse-redis-connection-in-socket-io

- Create the Message Hub service that will host our socket.io server. This is an example of a microservice in node https://github.com/hossein-85/WASP/tree/master/server
 - The way I envisaged the Message Hub service to work was that it would subscribe to a channel on redis
 - Then gateway receives the message via a rest route from the client and then will publish that to the channel that Message hub is listening to.
 - Once a message is received on that channel, Message Hub would relay the message to the recipients using socket.io. This is a bit different to the implementation that you will find in the examples I posted on the ticket as we do not have to subscribe to any socket.io channels (we subscribe to redis instead). If we do it our way, the Message Hub could only focus on relaying the messages and all the other work involved in sending and receiving a chat message can be handled by another service like a Chat service or something, it can do things like persisting messages in the DB, storing assets (images, audio, video, ..) 
in a repository and things like that.

## Good to have
- Documentation

- When building the message hub have scaling the service in mind. This article touches on a very important point that we need to take into consideration https://github.com/rajaraodv/redispubsub#socketio-and-sticky-sessions

- Also, it is very likely that we run into a problem with clients being able to register to the socket.io hub\server when start re-routing traffic from client through Gateway to the socket server. Test this and if it is a problem find a solution. This article might offer a solution https://github.com/rajaraodv/redispubsub#socketio-and-sticky-sessions
